### PR TITLE
Data section in tool.yml and parameters.json

### PR DESCRIPTION
**Pull request by AlexDo1**
This is a first suggestion on how we can adapt the tool specification to separate `parameters` from `data` with an additional section in `tool.yml` and `parameters.json`.

First idea for data fields are:
* `path`
* `format`
* `parse`
* `include`

The names of the fields and there meaning is of course up to discussion and we can remove fields or add additional fields.

I added the specification of `Data` to the file `parameter.md`, we can also think about adding a new file `data.md`, where data is defined separately from the parameters. As parameters and data are defined together in `tool.yml` and `parameters.json`, I thought that it can also make sense to keep the specification together.

### CODE REVIEW

File: docs/input.md [hunk 1]
```diff
@@ -97,9 +103,9 @@ tool:
           - option 3
 ```
 
-#### file
+#### `file`
 
-The `type=file` type indicates, that the passed parameter is a path to a file. These files contain data in most cases, but may also contain more complex configurations.
+The `type=file` type indicates, that the passed parameter is a path to a file. These files should not contain input data for the tool, as input data is specified in the `data` section of `tool.yml`. The type `file` can be used for more complex configurations that are stored in any kind of file format.
 The library used for parsing parameterizations should load the file into memory and pass the respective datastructure to the tool. This way, the tool is independent of specific data files. 
 If this is not a possible workflow, file paths can be passed as ordinary string parameters and the parsing library will not attempt to load the file.
```
Comment (mmaelicke) [inline docs/input.md:110]: this can be removed, as we decided that there is no `type=file` anymore

File: docs/input.md [hunk 2]
```diff
@@ -110,15 +116,15 @@ There are a number of file types, which are loaded by default:
 | .dat  |  `numpy.array` | `vector` | `matrix`  | `number[][]` | 
 | .csv  |  `pandas.DataFrame` | `data.frame` |  `matrix` |  `number[][]` |
 
-#### asset
+#### `asset`
 
 The `type=asset` can be used to specify paths to files or entire folders that are copied unchanged to the `/in/` path of the tool container and thus made available to the tool for further processing. The parsing library never attempts to load and process these files, therefore the files are available in the tool exactly as they are.  
-Assets can be tool configurations, folders containing data, geometry files or all kinds of other files.
+Assets can be tool configurations, geometry files or all kinds of other files.
```
Comment (mmaelicke) [inline docs/input.md:122]: "Assets are files that are available *as-is* in the container. Assets are neither Data nor parameters, but their dynamic nature might influence the tool execution. Hence, they are added as input to the tool."

File: docs/input.md [hunk 3]
```diff
@@ -129,29 +135,118 @@ description: |
...
+### Fields
+
+The following section defines all mandatory and optional fields of a `Data` entity.
+
+#### `path`
+
+The `path` field is the only mandatory field. Each input data needs a path, as input data
+is always given to the tool in a file- or folder-based way.
+Just as for paramters of `type=file`, the library used for parsing parameterizations and 
+data loads the file into memory and passes the respective datastructure to the tool, which
```
Comment (mmaelicke) [inline docs/input.md:189]: `type=file` should not exist anymore, after this revision

File: docs/input.md [hunk 4]
```diff
@@ -129,29 +135,118 @@ description: |
...
+passed just as strings, the parsing library will not attempt to load the file. 
+
+There are a number of file types, which are loaded by default:
+
+| file extension | Python |  R  |  Matlab |  NodeJS  |
+| ---------------|--------|-----|---------|----------| 
+| .dat  |  `numpy.array` | `vector` | `matrix`  | `number[][]` | 
+| .csv  |  `pandas.DataFrame` | `data.frame` |  `matrix` |  `number[][]` |
+
+#### `format`
```
Comment (mmaelicke) [inline docs/input.md:201]: hmm, format might be mis-leading here. What do you think about `mime` or `extension`?

File: docs/input.md [hunk 5]
```diff
@@ -129,29 +135,118 @@ description: |
...
+  foobar:
+    parameters:
+      ...
+    data:
+      foo_data:
+        path: /path/to/folder/containing/nc/files/
+        format: .nc
+```
+
+#### `parse`
```
Comment (mmaelicke) [inline docs/input.md:220]: Use `preload`, `load` or `lazy-load`. Parsing is not strictly correct

File: docs/input.md [hunk 6]
```diff
@@ -129,29 +135,118 @@ description: |
...
+and passed in the respective datastructure to the tool. Instead, file paths are passed as 
+ordinary strings and the parsing library will not attempt to load the file.
+
+#### `include`
+
+Boolean field which defaults to `true`. If set to `include=false`, the file / data is not copied 
+to the `/in/` folder of the container. Instead of this, the location of the locally saved file or
+folder is mounted to the docker container. Through this, hashsums of the data cannot be calculated
+and workflows cannot be proven to be reproducible anymore, but the tool execution can be accelerated
+and hard drive memory is saved, especially if input data is very large.
```
Comment (mmaelicke) [inline docs/input.md:233]: remove

File: docs/input.md [hunk 7]
```diff
@@ -180,4 +273,14 @@ tools:
...
         description: An optional array of floats
+    data:
+      foo_csv_data:
+        path: /path/to/data.csv
+        description: A .csv file containing input data
+      foo_folder_data:
+        path: /path/to/folder
+        format: .nc
+        parse: false
+        include: false
```
Comment (mmaelicke) [inline docs/input.md:284]: remove

File: docs/input.md [hunk 8]
```diff
@@ -1,18 +1,20 @@
...
+required along with specifications, e.g. about value ranges, default values and
+data types. The *parameterization* is file based and defaults to 
 `/in/parameters.json`. The JSON format is mandatory. 
 
 ## Missing parameterization
 
-In case a [Tool](./tool.md) accepts only optional parameters, or no parameters 
-are defined at all, the `/in/parameter.json` can be an empty file:
+In case a [Tool](./tool.md) accepts only optional parameters and no input data,
+or no parameters and data are defined at all, the `/in/parameter.json` can be
```
Comment (mmaelicke) [inline docs/input.md:16]: `/in/input.json`

File: docs/input.md [hunk 9]
```diff
@@ -26,25 +28,28 @@ In Python this would be an empty `dict`, in R an empty `list`.
...
 if the parameterization of a tool can be applied to other data. That means, the 
 **same** analysis is run on **different** data.
+Input data of a tool is also defined in *tool.yml* and *parameters.json*, but in
+a separate section `data`.
 
 From a practical perspective, if you build a tool around these tool specifications,
-the tool name and content of the `/in/parameter.json` can be used to create 
-checksums.
+the tool name and content of the sections `parameters` and `data` of the 
+`/in/parameter.json` can be used to create checksums and therefor help to establish
```
Comment (mmaelicke) [inline docs/input.md:48]: input.json

File: docs/input.md [hunk 10]
```diff
@@ -76,12 +82,12 @@ Allowed data-types include:
...
 * asset
 
-#### enum
+##### `enum`
 
 The `type=enum` field has an additional mandatory `values` field, which lists all
 allowed enum values. Note that enums should be validated by a parsing library
 or a library calling the tools. For the tools, enums parameters are treated like 
-strings as soon as read from a `parameter.json` file.
+strings as soon as read from a `parameters.json` file.
```
Comment (mmaelicke) [inline docs/input.md:90]: input.json

**Review by mmaelicke — CHANGES_REQUESTED** [review]
I found a few references to `parameters.json`, but you should double-check  everything again. 
There is also the `type=file` section still in there, which should be removed.

File: docs/input.md (hunk 4 above)
Comment (AlexDo1) [reply docs/input.md]: I like `extension`.

**Comment by AlexDo1** [comment]
Thanks for the review, it took a little bit longer until I finished everything today.

Ready for review, I can also have a look again in the next days.

**Comment by AlexDo1** [comment]
@mmaelicke this is the PR about updating tool-specs and tool.

**Pull request merged**